LightOj/Graph Theory/1002 - Country Roads.txt

```diff
@@ -0,0 +1,142 @@
+/*
+Algorithm: BFS + Kruskal 
+
+Explanation: It is obvious that we need to find the Minimum Spanning Tree(MST) first 
+so as to eliminate maximum weight edges.
+Then in this MST we can use BFS to keep track of maximum edge to a vertex in each level.
+ 
+*/
+#include<bits/stdc++.h>
+using namespace std;
+#define opt ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL)
+#define lli long long int
+#define mod		1000000007
+#define inf     1000000000000000000
+#define MX1		100005
+#define MX2		10000004
+#define pi		acos(-1)
+#define pb      push_back
+#define vlli    vector<lli>
+#define plli    vector<pair<lli,lli> >
+#define mp      make_pair
+#define all(v)  v.begin(),v.end()
+#define test    int t;cin>>t;
+lli gcd(lli a, lli b){return b?gcd(b,a%b):a;}
+lli power(lli a,lli b){lli ans=1;while(b){if(b&1){ans=(ans*a)%mod;}a=(a*a)%mod;b>>=1;}return ans;}
+plli tr[505];
+bool visit[505];
+lli dist[505];
+lli id[505];
+lli n,m;
+void initialize()
+{
+	for(lli i=0;i<505;i++)
+	id[i]=i;
+}
+lli root(lli x)
+{
+	while(x!=id[x])
+	{
+		id[x]=id[id[x]];
+		x=id[x];
+	}
+	return x;
+}
+void union1(lli x,lli y)

```

Comment:
Since union is a keyword in c++ language,it is not allowed to make it as a function function.So i made it as union1. If there is any problem i will change it as "merge".